Leverage virtualenv to ensure pip and setuptools and wheel are recent

jaraco opened the pull request:
Fixes #24. Obviates #23. Fixes #17.

jaraco committed:
Leverage virtualenv to ensure pip and setuptools and wheel are recent. Fixes #24.

jaraco commented:
Try out this PR here: https://m.devpi.net/jaraco/dev/tox-venv/0.3.1+pr25

jaraco commented:
I'll be honest, I don't fully understand the test failures.

Some of them seem spurious to me, like the assertions are too tightly coupled to the implementation details.

Other ones may be more legitimate, revealing that in some cases 'python' doesn't exist. I'm not sure why that would be.

So I present this PR for your consideration. It works for me and it solves the issue in my environment. I don't do a lot of Python 3.5 work, but when I do, it's nice to have this functionality.

asottile commented:
A few minor concerns, it's probably fine:

- does this potentially _downgrade_ packages?  For instance say I'm running virtualenv 15.0.1 w/ python 3.7.0
    - 3.7.0 ships with pip 10.0.1, setuptools 39.0.1
    - virtualenv 15.0.1 ships with pip 8.1.1, setuptools 20.3 (and wheel and argparse)
- does this work for debian's mangled `pip` / `virtualenv` packages? (probably ok to just not, but probably worth considering)

jaraco commented:
> does this potentially _downgrade_ packages?

I just installed virtualenv 15.0.1 and ran `tox -r` on a Python 3.7 environment (with this patched tox-venv installed), and pip 10.0.1 was still present in the new env. So it seems it does not downgrade packages.

> does this work for debian's mangled pip / virtualenv packages?

Hard to say, but I'd imagine so, in part because it's leveraging the virtualenv code, but also because I just ran a test with this branch on ubuntu bionic... and it creates the environments just fine. Is there a test I could run that would reveal an anticipated failure?

asottile reviewed (approved):
probably sufficient then!

jaraco commented:
Thanks. In that case, I'll try to get the tests to pass.

jaraco commented:
I'm having difficulty running the tests. I can't get the tests on master to pass, nevermind the patch. I get 18 failures, most similar to those failing in Travis. Can someone help with ferreting out the issues with the tests failing - maybe give some advice how I might proceed?

rpkilby commented:
Hi @jaraco. I haven't had time to look at this yet, but my immediate guess is that the test failures are due to the test suite being out of date. One maintenance annoyance is making sure the test suite is kept up to date, the reapplying the tox-venv related changes. 

Try temporarily pinning `tox==3.0.0`, as that was the last test suite update.

I'll try to take a look at this soon.

jaraco commented:
Cool. I created the [stable-deps branch](https://github.com/jaraco/tox-venv/tree/stable-deps), where tests are passing for me. I then rebased the work in this branch on there in the [stable-deps+feature/update-env-installer branch](https://github.com/jaraco/tox-venv/tree/stable-deps%2Bfeature/update-env-installer), and many tests are failing, revealing apparent regressions. At least the concerns are separated now.

jaraco commented:
[These errors](https://travis-ci.org/jaraco/tox-venv/jobs/431652258) match what I'm seeing locally. Probably the mocks need to be updated to mock out the virtualenv install_wheel call.

jaraco commented:
I've pushed 8b59f63e975, and I expect the tests to pass in Travis for that branch, but it doesn't prove much; it just disables the added functionality to prevent it from causing the tests to fail. 
At this point, I need to defer to someone with more understanding of the mock technique to decide how's best to test (or not test) this functionality.

jaraco commented:
In pypa/setuptools#1500, I discovered a regression. It seems that when testing setuptools, including this functionality causes the `develop`-installed setuptools to be replaced by this vendored copy. I'll want to investigate why that is and why a virtualenv isn't affected.

jaraco committed:
Use 'str' to convert a LocalPath to a value suitable for install_wheel. Fixes test failures on Python 3.5

jaraco commented:
Today I found another regression elicited by the test suite on Python 3.5 (and probably Python 2.7). That's fixed in d988843. And as you can see in [this build](https://travis-ci.org/jaraco/tox-venv/builds/473406854), if the dependencies are pinned to avoid unrelated issues in the dependencies, the tests pass on the same Pythons in which [master passes with those some pins](https://travis-ci.org/jaraco/tox-venv/builds/431653343).

I'd like to propose merging this PR (and cutting an 0.4 release). I've used this technique for months with no issues other than the apparent regression I reported on Sep 23 that hasn't re-emerged, and I again today confirmed it's not happening in a typical environment.

jaraco commented:
Hmm. I may have run into a new issue with this code. This morning, I've created literally hundreds of virtualenvs using this tox-venv version. But now I'm trying it and it's failing with this traceback:

```
jaraco.financial master $ tox                                                                                                                                                            
python create: /Users/jaraco/code/main/jaraco.financial/.tox/python
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.7/bin/tox", line 11, in <module>
    sys.exit(cmdline())
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/tox/session.py", line 42, in cmdline
    main(args)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/tox/session.py", line 49, in main
    retcode = build_session(config).runcommand()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/tox/session.py", line 457, in runcommand
    return self.subcommand_test()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/tox/session.py", line 583, in subcommand_test
    if self.setupenv(venv):
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/tox/session.py", line 508, in setupenv
    status = venv.update(action=action)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/tox/venv.py", line 217, in update
    self.hook.tox_testenv_create(action=action, venv=self)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/pluggy/__init__.py", line 617, in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/pluggy/__init__.py", line 222, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/pluggy/__init__.py", line 216, in <lambda>
    firstresult=hook.spec_opts.get('firstresult'),
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/pluggy/callers.py", line 201, in _multicall
    return outcome.get_result()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/pluggy/callers.py", line 76, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/pluggy/callers.py", line 180, in _multicall
    res = hook_impl.function(*args)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/tox_venv/hooks.py", line 111, in tox_testenv_create
    py_executable=os.path.join(venv.path / exe_dir / 'python'),
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/virtualenv.py", line 904, in install_wheel
    call_subprocess(cmd, show_stdout=False, extra_env=env, stdin=SCRIPT)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/virtualenv.py", line 796, in call_subprocess
    % (cmd_desc, proc.returncode))
OSError: Command /Users/jaraco/code/m...ox/python/bin/python - setuptools pip wheel failed with error code 2
```

Immediately I notice one deficiency - if an issue does occur when venv calls virtualenv to upgrade the packages, the output is [unconditionally suppressed](https://github.com/pypa/virtualenv/blob/5f59fcf42acc67aefbe970e45cef253644b32168/virtualenv.py#L983). Let's see if I can find a way to override that setting to see what the underlying error is.

jaraco commented:
Aha - so while trying to create a breakpoint in the stack trace above, I found it was failing earlier, and from there I was able to ascertain the underlying cause:

```
jaraco.financial master $ .tox/python/bin/python3 -Im ensurepip --upgrade --default-pip                                                                                                  
Looking in links: /var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/tmpqdilxrv1
Collecting setuptools
Collecting pip
Installing collected packages: setuptools, pip
Exception:
Traceback (most recent call last):
  File "/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/tmpqdilxrv1/pip-18.1-py2.py3-none-any.whl/pip/_internal/cli/base_command.py", line 143, in main
    status = self.run(options, args)
  File "/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/tmpqdilxrv1/pip-18.1-py2.py3-none-any.whl/pip/_internal/commands/install.py", line 366, in run
    use_user_site=options.use_user_site,
  File "/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/tmpqdilxrv1/pip-18.1-py2.py3-none-any.whl/pip/_internal/req/__init__.py", line 49, in install_given_reqs
    **kwargs
  File "/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/tmpqdilxrv1/pip-18.1-py2.py3-none-any.whl/pip/_internal/req/req_install.py", line 760, in install
    use_user_site=use_user_site, pycompile=pycompile,
  File "/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/tmpqdilxrv1/pip-18.1-py2.py3-none-any.whl/pip/_internal/req/req_install.py", line 382, in move_wheel_files
    warn_script_location=warn_script_location,
  File "/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/tmpqdilxrv1/pip-18.1-py2.py3-none-any.whl/pip/_internal/wheel.py", line 215, in move_wheel_files
    prefix=prefix,
  File "/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/tmpqdilxrv1/pip-18.1-py2.py3-none-any.whl/pip/_internal/locations.py", line 153, in distutils_scheme
    d.parse_config_files()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/distutils/dist.py", line 406, in parse_config_files
    parser.read(filename)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/configparser.py", line 696, in read
    self._read(fp, filename)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/configparser.py", line 1110, in _read
    raise e
configparser.ParsingError: Source contains parsing errors: 'setup.cfg'
        [line 57]: 'console_scripts\n'
```

I had a malformed setup.cfg... and that's almost certainly what's tripping up the invocation. Probably it would  be best if the invocation of `virtualenv.install_wheel` happened in a temporary directory so as to avoid any influence of a setup.cfg in the current working directory.

jaraco commented:
> virtualenv 16.2 was released this morning, perhaps related?

I was thinking the same thing, but I'm still running the older version.

asottile commented:
ahah, yeah I've hit that in the past too -- pre-commit works around this by invoking in `/`: https://github.com/pre-commit/pre-commit/blob/b31cb22d8b3fc2beb24ab451f5e1ac0121c95ee1/pre_commit/languages/python.py#L153

jaraco commented:
I've confirmed that the `setup.cfg` issue is a known issue in virtualenv, so not something that can or should be solved here. Ideally the tests should be passing, but I'm fairly confident the failures are outside the scope of this effort. I'm hoping the PR can be merged as-is or that someone can address the test failures. Thanks all for the feedback.

rpkilby commented:
Hi @jaraco. I've fixed a few issues with the test suite and updated it to reflect tox 3.8.1
Do you think you could take a second look at this, based on the latests changes.

jaraco committed:
Don't bother asserting that output is not empty. Many commands, including 'python -m venv', don't emit any output. Fixes #38.

jaraco committed:
Merge branch 'bugfix/38-no-assert-output' into feature/update-env-installer

jaraco committed:
Extract function for updating installers

jaraco commented:
The same issue persists. Many (15) of the tests fail when `install_wheel` tries to update a virtualenv that doesn't exist. I can elicit failures in the same tests by making this change:

```diff
diff --git a/src/tox_venv/hooks.py b/src/tox_venv/hooks.py
index 543fab9..5167168 100644
--- a/src/tox_venv/hooks.py
+++ b/src/tox_venv/hooks.py
@@ -104,5 +104,6 @@ def tox_testenv_create(venv, action):
     basepath.ensure(dir=1)
     args.append(venv.path.basename)
     venv._pcall(args, venv=False, action=action, cwd=basepath)
+    assert venv.path.isdir()
     # Return non-None to indicate the plugin has completed
     return True
```

Under what conditions is it reasonable for that `venv._pcall` to be invoked but no directory created?

jaraco committed:
Only invoke install_wheel when the virtualenv exe directory exists. Fixes failing tests.

jaraco committed:
Update changelog

jaraco committed:
Run isort

jaraco commented:
This last commit should pass the tests. Ultimately, 991ad52 was the change that was needed to avoid the test failures. I believe this latest implementation is now more robust and ready for acceptance.

Note that this branch does depend on and incorporate the fix in #39, so if you accept this PR, it will also close #39.

jaraco commented:
[Preview this code](https://m.devpi.net/jaraco/dev/tox-venv/0.4.0+pr25) against the latest 0.4 release with:

```
pip install -i https://m.devpi.net/jaraco/dev tox-venv==0.4.0+pr25
```

rpkilby commented:
Thanks for the update - I'll have a look sometime this week.

jaraco commented:
@rpkilby I'm hoping you'll have a chance to review/accept this and cut a release with it.

merwok commented on CHANGELOG.rst:
```diff
@@ -1,6 +1,9 @@
 ==========
 
+- #17: Like with ``virtualenv``, venvs are always created with a
+  late version of pip, setuptools, and wheel.
```
«a late version»? does that mean: the latest version?

jaraco replied:
I'm of the sense that sometimes the latest version is not installed. In fact, just now I created a venv using tox-venv with this patch, and it got pip 19.0.3 not 19.1:

```
pip-run master $ tox -r --notest                                                                                                                                               
python recreate: /Users/jaraco/code/main/pip-run/.tox/python
python installdeps: setuptools>=31.0.1
python develop-inst: /Users/jaraco/code/main/pip-run
python installed: atomicwrites==1.3.0,attrs==19.1.0,docutils==0.14,entrypoints==0.3,flake8==3.7.7,importlib-metadata==0.9,mccabe==0.6.1,more-itertools==7.0.0,-e git+gh://jaraco/pip-run@b92009576f96c4e02cf7f22db6db393ee18d3fbe#egg=pip_run,pluggy==0.9.0,py==1.8.0,pycodestyle==2.5.0,pyflakes==2.1.1,pytest==4.4.1,pytest-checkdocs==1.2.0,pytest-flake8==1.0.4,six==1.12.0,zipp==0.3.3
___________________________________________________________________________________ summary ___________________________________________________________________________________
  python: skipped tests
  congratulations :)

pip-run master $ .tox/python/bin/pip --version                                                                                                                                 
pip 19.0.3 from /Users/jaraco/code/main/pip-run/.tox/python/lib/python3.7/site-packages/pip (python 3.7)
pip-run master $ which tox                                                                                                                                                     
/Users/jaraco/.local/bin/tox
pip-run master $ ~/.local/pipx/venvs/tox/bin/pip show tox-venv                                                                                                                 
Name: tox-venv
Version: 0.4.0+pr25
Summary: Use Python 3 venvs for Python 3 tox testenvs
Home-page: https://github.com/tox-dev/tox-venv
Author: Ryan P Kilby
Author-email: kilbyr@gmail.com
License: BSD
Location: /Users/jaraco/.local/pipx/venvs/tox/lib/python3.7/site-packages
Requires: tox
Required-by: 
```

So I don't think it's always the latest version.

Weirdly, I end up with a different version when using virtualenv:

```
pip-run master $ ~/.local/pipx/venvs/tox/bin/python -m virtualenv venv                                                                                                         
Using base prefix '/Library/Frameworks/Python.framework/Versions/3.7'
New python executable in /Users/jaraco/code/main/pip-run/venv/bin/python
Installing setuptools, pip, wheel...
done.
pip-run master $ venv/bin/pip --version                                                                                                                                        
pip 19.1 from /Users/jaraco/code/main/pip-run/venv/lib/python3.7/site-packages/pip (python 3.7)
```

This finding is indeed disconcerting.

jaraco replied:
I think to get _equivalent_ behavior (the intention of this PR), it may be necessary to invoke the [virtualenv_support_dirs context](https://github.com/pypa/virtualenv/blob/debbaa4ff2712167006f21d978267968c2191d2b/virtualenv.py#L474).

jaraco replied:
Hmm. Now that I look at it again, I wonder if this patch is having any effect at all. At one point, I'm certain it was. But right now, I'm seeing the same version of pip appearing from vanilla `python -m venv` as from `tox` plus `tox-venv+pr25`.

jaraco replied:
I see now that [virtualenv_support_dirs is the default](https://github.com/pypa/virtualenv/blob/debbaa4ff2712167006f21d978267968c2191d2b/virtualenv.py#L968). Probably the difference I'm seeing in behavior is the fact that [download defaults to True](https://github.com/pypa/virtualenv/blob/debbaa4ff2712167006f21d978267968c2191d2b/virtualenv.py#L692) for the command-line invocation but [defaults to False](https://github.com/pypa/virtualenv/blob/debbaa4ff2712167006f21d978267968c2191d2b/virtualenv.py#L966) for the API invocation.

jaraco replied:
In this latest commit (7c0f2b5), I pass the `download=True` to the `install_wheel`, which causes the install to download the latest versions, same as virtualenv does when invoked by tox without venv, and so I've updated the changelog accordingly.

Also, I verified the behavior against [another pre-release](https://m.devpi.net/jaraco/dev/tox-venv/0.4.0+pr25.2):

```
tox-venv feature/update-env-installer $ ~/.local/pipx/venvs/tox/bin/python -m pip install -i https://m.devpi.net/jaraco/dev -U tox-venv                                        
Looking in indexes: https://m.devpi.net/jaraco/dev
Collecting tox-venv
  Downloading https://m.devpi.net/jaraco/dev/+f/dae/1dff6d9bc24bf/tox-venv-0.4.0+pr25.2.tar.gz
Requirement already satisfied, skipping upgrade: tox>=3.8.1 in /Users/jaraco/.local/pipx/venvs/tox/lib/python3.7/site-packages (from tox-venv) (3.8.6)
Requirement already satisfied, skipping upgrade: filelock<4,>=3.0.0 in /Users/jaraco/.local/pipx/venvs/tox/lib/python3.7/site-packages (from tox>=3.8.1->tox-venv) (3.0.10)
Requirement already satisfied, skipping upgrade: six<2,>=1.0.0 in /Users/jaraco/.local/pipx/venvs/tox/lib/python3.7/site-packages (from tox>=3.8.1->tox-venv) (1.12.0)
Requirement already satisfied, skipping upgrade: py<2,>=1.4.17 in /Users/jaraco/.local/pipx/venvs/tox/lib/python3.7/site-packages (from tox>=3.8.1->tox-venv) (1.8.0)
Requirement already satisfied, skipping upgrade: virtualenv>=1.11.2 in /Users/jaraco/.local/pipx/venvs/tox/lib/python3.7/site-packages (from tox>=3.8.1->tox-venv) (16.4.3)
Requirement already satisfied, skipping upgrade: setuptools>=30.0.0 in /Users/jaraco/.local/pipx/venvs/tox/lib/python3.7/site-packages (from tox>=3.8.1->tox-venv) (40.8.0)
Requirement already satisfied, skipping upgrade: toml>=0.9.4 in /Users/jaraco/.local/pipx/venvs/tox/lib/python3.7/site-packages (from tox>=3.8.1->tox-venv) (0.10.0)
Requirement already satisfied, skipping upgrade: pluggy<1,>=0.3.0 in /Users/jaraco/.local/pipx/venvs/tox/lib/python3.7/site-packages (from tox>=3.8.1->tox-venv) (0.9.0)
Installing collected packages: tox-venv
  Found existing installation: tox-venv 0.4.0+pr25
    Uninstalling tox-venv-0.4.0+pr25:
      Successfully uninstalled tox-venv-0.4.0+pr25
  Running setup.py install for tox-venv ... done
Successfully installed tox-venv-0.4.0+pr25.2
tox-venv feature/update-env-installer $ tox --notest                                                                                                                           
GLOB sdist-make: /Users/jaraco/code/public/tox-venv/setup.py
python create: /Users/jaraco/code/public/tox-venv/.tox/python
python installdeps: tox[testing]>=3.8.1
python inst: /Users/jaraco/code/public/tox-venv/.tox/.tmp/package/2/tox-venv-0.4.0.zip
python installed: apipkg==1.5,atomicwrites==1.3.0,attrs==19.1.0,coverage==4.5.3,execnet==1.6.0,filelock==3.0.10,flaky==3.5.3,freezegun==0.3.11,more-itertools==7.0.0,pathlib2==2.3.3,pluggy==0.9.0,psutil==5.6.1,py==1.8.0,pytest==4.4.1,pytest-cov==2.6.1,pytest-forked==1.0.2,pytest-mock==1.10.4,pytest-randomly==1.2.3,pytest-xdist==1.28.0,python-dateutil==2.8.0,six==1.12.0,toml==0.10.0,tox==3.9.0,tox-venv==0.4.0,virtualenv==16.5.0
___________________________________________________________________________________ summary ___________________________________________________________________________________
  python: skipped tests
  congratulations :)

tox-venv feature/update-env-installer $ .tox/python/bin/pip --version                                                                                                          
pip 19.1 from /Users/jaraco/code/public/tox-venv/.tox/python/lib/python3.7/site-packages/pip (python 3.7)
```

jaraco committed:
When installing wheel, use download=True, same as default virtualenv behavior.

jayvdb commented:
I expect this is going to break a lot of people's use of tox-venv, as they expect that with this plugin the new venv is raw from Python, which is very stable except that the builtin Python `venv` already upgrades pip.

I dont understand why people would want to use tox-venv if this gets merged, as it seems easier to just use the standard tox without tox-venv which will do all this.

jaraco commented:
> the builtin Python `venv` already upgrades pip.

That's not my experience:

```
$ python -m venv test                                                                                                   
$ test/bin/python -m pip --version                                                                                      
pip 19.0.3 from /Users/jaraco/code/public/pypa/twine/test/lib/python3.7/site-packages/pip (python 3.7)
```

Note that pip 19.1.1 is the latest pip at this time of writing.

> I dont understand why people would want to use tox-venv if this gets merged, as it seems easier to just use the standard tox without tox-venv which will do all this.

There are many reasons.

- venv addresses issues inherent in the design of virtualenv that make it better behaved and have fewer bugs (I don't have a reference handy, but the references in the original post may illustrate some).
- venv is newer than virtualenv and intends to supersede virtualenv for later Pythons; it's the future, so early adopters should be able to opt-in
- venv is more robust against changes to the interpreter
- venv is built-in, so has no variability in versions and is thus more predictable
- venv is lighter weight, having no dependencies

All of these reasons are why I use `tox-venv` (in roughly descending order of importance). My understanding also is that tox intends to adopt this functionality by default, not as a plugin, so having it operate at parity with the existing virtualenv implementation is an important step in that transition.

jayvdb commented:
Builtin venv has option `with_pip` which can be used to upgrade pip, if that is the main goal of this PR, without using `virtualenv`.   It would be great if `tox-venv` exposed that `venv` options as tox configuration, enabled by default.

As noted elsewhere, mandatory installation and upgrading of `wheel` is weird and should be not copied.

That only leaves the problem of auto upgrading setuptools, which is a simple problem to solve without `virtualenv` if `venv` has already upgraded `pip`.

asottile commented:
> Builtin venv has option `with_pip` which can be used to upgrade pip, if that is the main goal of this PR, without using `virtualenv`. It would be great if `tox-venv` exposed that `venv` options as tox configuration, enabled by default.

I only see `--without-pip` which prevents installation of pip into the resulting virtualenv (and I'm not sure why that option would eve be used) -- as far as I can tell `with_pip` (from the python api) only ensures that `ensurepip` is called on the final virtualenv

asottile commented:
and `ensurepip` uses the wheels bundled with the interpreter -- it does not upgrade from pypi

jayvdb commented:
https://github.com/python/cpython/blob/master/Lib/venv/__init__.py#L269 - it uses --upgrade

https://github.com/python/cpython/blob/master/Lib/ensurepip/__init__.py#L110 passes upgrade to pip

asottile commented:
You missed the `--no-index -find-links` directly above :wink: 

https://github.com/python/cpython/blob/ff6bb0aa95259413f359d42410526ff0b4dccfb7/Lib/ensurepip/__init__.py#L107

jayvdb commented:
Bugger!  I guess they also decided to choose stability over dephell of constantly updating pip/setuptools/etc., a reason I used tox-venv.  As we can disregard 2.7, 3.4, etc now all EOL or close to EOL, most people are using venv bundles from py36, pip-18.1 and setuptools-40.6.2 .

https://github.com/python/cpython/tree/3.6/Lib/ensurepip/_bundled

And they are being updated fairly frequently

https://github.com/python/cpython/commits/master/Lib/ensurepip/_bundled

We're not in the days of distros providing pip 0.3-ish, which couldnt even read 90% of requirements files in the wild.

For people needing more recent versions, another approach is tox-pip-version , which does provide upgrading, as this is also a feature requested of tox - https://github.com/tox-dev/tox/issues/791 .  The beauty of plugins where each does one thing.

rpkilby commented:
Leaving this functionality to tox-pip-version seems sensible to me. @jaraco any reason this wouldn't work?

jaraco commented:
> any reason this wouldn't work?

I have no reason to expect that `tox-pip-version` can't be used to specify a pip version... and other features could be used to specify setuptools/wheel versions.

I still think you want this change to get parity between virtualenv and env envs (even for just between py2/3 when tox-venv is present). Another way to achieve parity would be to instead use ensurepip in virtualenv envs.

jaraco commented:
II agree it seems nicer to separate the concerns of pip version requirements and virtualenv management. I'll give tox-pip-version a try and stop fighting this battle.

rpkilby commented:
Apologies - never meant for this to be difficult. I'm just stretched on time. Currently, I'm waiting on the tox 4 to see if this is even necessary. But for now, using tox-pip-version is the way to go.

Aricg commented:
tox-pip-version does not seem to work with tox-venv
one uses virtualenv and one uses python -m venv
am I missing something?

How can I at the same time not use virtualenv with tox and also have a reasonable version of pip?

Aricg commented:
I take it back! it works!

`16:27:55 coala run-test: commands[0] | python check-pip-version.py 18.1
16:27:56 PASS: version '18.1' is in set '==18.1'`

Aricg commented:
Thanks for making this package. tox-venv and tox-pip-version have combined to saved me from the Fifth Circle of python hell.



`[testenv:coala]
basepython = python3
pip_version = pip==18.1
deps =
    coala
    coala-bears
    pygments~=2.3.1
    requests~=2.21.0
    nodeenv
    numpy~=1.16.4
commands =
    python check-pip-version.py 18.1
    nodeenv -p
    npm install --global remark-cli remark-lint write-good
    coala --non-interactive
`

jaraco commented:
Unfortunately, it seems that `tox-pip-version` is not suitable for the case of a bootstrap environment (`.tox/.tox`). See pypa/setuptools#2000.

rpkilby commented:
Hi @gaborbernat. Given the changes to virtualenv 20, is there a "recommended" but unofficial way of installing the bundled wheels? Right now, it looks like we'd need to do something like:

```python
from virtualenv.run import session_via_cli

session = session_via_cli(['path/to/venv', '--seeder', 'pip'])
session.seeder.run(session.creator)
```

gaborbernat commented:
Out of box it installs bundled wheels.

jaraco commented:
venv is dead. long live virtualenv.
